### PR TITLE
fix: Prevent metadata duplication

### PR DESCRIPTION
**Pull request by Jimlinz**
Pystac `normalize_hrefs` resolves all links and mutates the entire tree, causing `<item>.json` and `collection.json` for all datasets within the catalog to update. `save_object` only saves the particular entity within the catalog, preventing unnecessary metadata duplication. 

## Reference
https://github.com/stac-utils/pystac/issues/284
https://github.com/stac-utils/pystac/pull/294
https://github.com/stac-utils/pystac/issues/90

### CODE REVIEW

File: geostore/populate_catalog/task.py [hunk 1]
```diff
@@ -95,8 +95,4 @@ def handle_message(metadata_key: str) -> None:
     if root_catalog.get_child(dataset_metadata.id) is None:
         root_catalog.add_child(child=dataset_metadata, strategy=GeostoreSTACLayoutStrategy())
 
-    root_catalog.normalize_hrefs(
```
Comment (billgeo) [inline geostore/populate_catalog/task.py:98]: We do want to normalize hrefs across the Geostore (hopefully we have a test to enforce that). Have you checked that we have code and/or a test for that?

Comment (billgeo) [reply geostore/populate_catalog/task.py:98]: We definitely don't need to normalize every href every time we add a dataset though, just need to do the catalog.json and all the STAC file in the child that is being added. Can ignore all the siblings.

**Comment by Jimlinz** [comment]
Changing PR to _draft_ since tests are failing. `save_object` appears to be a [suggestion](https://github.com/stac-utils/pystac/issues/90#issuecomment-614673898) to work around #1818; however, this is causing tests to fail. More work is needed to figure out why the tests are failing and if there is a better solution for the problem above.

File: geostore/populate_catalog/task.py (hunk 1 above)
Comment (Jimlinz) [reply geostore/populate_catalog/task.py:98]: This was addressed in https://github.com/stac-utils/pystac/pull/294 (if I read the PR correctly). Nevertheless, tests are failing which means more work is needed to figure out what is needed to fix this issue.

**Comment by l0b0** [comment]
Looking through the test failures:

- `should_update_root_catalog_with_new_version_catalog [dataset version links]` fails because an [optional `title`](https://github.com/radiantearth/stac-spec/blob/789e90fba4c7119bc925d26f75f47bacf44c91e6/catalog-spec/catalog-spec.md#link-object) is missing. Possible solutions:
   - Ignore the irrelevant field in the test. This is probably the simplest, but means the test might have to change again if PySTAC changes.
   - Change the test to only check the URLs rather than the whole `link` array. More robust, but also more work and less coverage.
- `should_update_root_catalog_with_new_version_collection [item links]` fails for two separate reasons:
   - Missing `title`, as above.
   - **Wrong root link.** This is an actual issue, and needs fixing.
- `should_successfully_run_dataset_version_creation_process_and_again_with_partial_upload` fails because of a **missing/wrong root link** (`expected_link_object` is never `in (load(smart_open.open(f"{storage_bucket_prefix}{dataset_title}/{item_metadata_filename}", mode="rb")))[STAC_LINKS_KEY]`). Also needs fixing.

In summary, to make this work ASAP we could:

1. Modify tests to remove the `title` from the relevant expected `links` entry.
2. Change the code which writes metadata files (`geostore.import_metadata_file.task.importer`) to modify the root URL as it's currently specified in the tests.

File: tests/test_populate_catalog.py [hunk 2]
```diff
@@ -269,15 +265,23 @@ def should_update_root_catalog_with_new_version_catalog(subtests: SubTests) -> N
             mode="rb",
         ) as updated_root_metadata_file:
             catalog_json = load(updated_root_metadata_file)
-            assert catalog_json[STAC_LINKS_KEY] == expected_root_catalog_links
+            updated_catalog_json = []
+            for i in catalog_json[STAC_LINKS_KEY]:
+                del i["title"]
```
Comment (l0b0) [inline tests/test_populate_catalog.py:270]: Rather than deleting the title from the actual state you could simply remove the non-existent title from the expected state. That should result in less code overall, and very slightly more coverage.

File: geostore/populate_catalog/task.py [hunk 3]
```diff
@@ -95,8 +95,4 @@ def handle_message(metadata_key: str) -> None:
...
-        f"{S3_URL_PREFIX}{Resource.STORAGE_BUCKET_NAME.resource_name}",
-        strategy=GeostoreSTACLayoutStrategy(),
-    )
```
Comment (l0b0) [inline geostore/populate_catalog/task.py:101]: Isn't there some missing code to actually replace the root `href`?

```
>               assert item_json[STAC_LINKS_KEY] == expected_item_links
E               AssertionError: assert [{'href': './CCRrWAn9FPPOyNFd1Q6I.json',\n  'rel': 'root',\n  'type': 'application/json'},\n {'href': './CCRrWAn9FPPOyNFd1Q6I.json',\n  'rel': 'parent',\n  'type': 'application/json'}] == [{'href': '../catalog.json', 'rel': 'root', 'type': 'application/json'},\n {'href': './CCRrWAn9FPPOyNFd1Q6I.json',\n  'rel': 'parent',\n  'type': 'application/json'}]
E                 At index 0 diff: {'rel': 'root', 'href': './CCRrWAn9FPPOyNFd1Q6I.json', 'type': 'application/json'} != {'rel': 'root', 'href': '../catalog.json', 'type': 'application/json'}
E                 Full diff:
E                   [
E                 -  {'href': '../catalog.json',
E                 +  {'href': './CCRrWAn9FPPOyNFd1Q6I.json',
E                     'rel': 'root',
E                     'type': 'application/json'},
E                    {'href': './CCRrWAn9FPPOyNFd1Q6I.json',
E                     'rel': 'parent',
E                     'type': 'application/json'},
E                   ]
```

File: tests/test_populate_catalog.py (hunk 2 above)
Comment (Jimlinz) [reply tests/test_populate_catalog.py]: Do you mean something like this https://github.com/linz/geostore/commit/bb4fd6295f1b8d88c724874e82bd5b198bf2d37e?
It was my initial attempt, but it didn't work. Maybe I have missed something? 

`del` didn't work, had to change it to `pop`

Comment (l0b0) [reply tests/test_populate_catalog.py]: Something like bb4fd62, except some of the titles should still be there.

File: geostore/populate_catalog/task.py [hunk 4]
```diff
@@ -63,7 +63,7 @@ def get_collection_href(self, col: Collection, parent_dir: str, is_root: bool) -
         assert not is_root
         return str(col.get_self_href())
 
-    def get_item_href(self, item: Item, parent_dir: str) -> str:
+    def get_item_href(self, item: Item, parent_dir: str) -> str:  # pragma: no cover
```
Comment (Jimlinz) [inline geostore/populate_catalog/task.py:66]: `GeostoreSTACStrategy` `get_item_href` is not called after `normalized_href` is removed; however, this looks like it may be relevant again in the future as the codebase grows (and keeping this for completeness). Excluding this function from coverage, for now.

File: geostore/import_metadata_file/task.py [hunk 5]
```diff
@@ -16,6 +16,7 @@
 from ..types import JsonObject
 
 S3_BODY_KEY = "Body"
+CATALOG_FILENAME = "catalog.json"
```
Comment (l0b0) [inline geostore/import_metadata_file/task.py:19]: This constant already exists in `geostore/populate_catalog/task.py`, so it should probably be moved to a file in the `geostore` directory and imported.

File: geostore/populate_catalog/task.py (hunk 4 above)
Comment (l0b0) [reply geostore/populate_catalog/task.py:66]: I would not leave dead code in place for "completeness". We'll have the repo history, so it can be recovered if we ever need it. Most likely we never will.

File: tests/test_processing_stack.py [hunk 6]
```diff
@@ -483,10 +511,9 @@ def should_successfully_run_dataset_version_creation_process_and_again_with_part
 
             #  Wait for Item metadata file to include a link to geostore root catalog
             expected_link_object = {
-                STAC_HREF_KEY: f"../{CATALOG_FILENAME}",
+                STAC_HREF_KEY: CATALOG_FILENAME,
```
Comment (l0b0) [inline tests/test_processing_stack.py:514]: I don't think this is right. The root should always be the actual root catalogue file, which is in the parent directory of the dataset.

File: geostore/populate_catalog/task.py (hunk 4 above)
Comment (Jimlinz) [reply geostore/populate_catalog/task.py:66]: It looks like `GeostoreSTACLayoutStrategy` overrides `HrefLayoutStrategy`, and can only be instantiated if it overrides all the abstract methods in the parent class??

File: geostore/import_metadata_file/task.py (hunk 5 above)
Comment (Jimlinz) [reply geostore/import_metadata_file/task.py]: Have changed this to reference `geostore/populate_catalog/task.py` (for now). Good idea to move this to another file, but a quick look shows it is reference in quite a few places, so I might move this change to another PR.

File: tests/test_processing_stack.py (hunk 6 above)
Comment (Jimlinz) [reply tests/test_processing_stack.py]: You are correct about this. Thanks for flagging it. The problem was caused by calling `update_root_link` before another function which `change_href_to_basename`. Took some time to figure out, but makes more sense now.

File: geostore/import_metadata_file/task.py (hunk 5 above)
Comment (Jimlinz) [reply geostore/import_metadata_file/task.py]: Spoke too soon. The value needs to be accessible via the lambda function. Changing this back to `../catalog.json` for now and will create another PR to address the above.

File: geostore/populate_catalog/task.py (hunk 3 above)
Comment (Jimlinz) [reply geostore/populate_catalog/task.py:101]: Took a while for me to trace this down. There are two options to tackle this bug: 
`root_catalog.save_object()` or `root_catalog.save(catalog_type=CatalogType.SELF_CONTAINED)` see `populate_catalog/task.py`

The first option `save_object` is the workaround suggested by Pystac (see https://github.com/stac-utils/pystac/issues/284). This would be the ideal implementation as none of the metadata gets duplicated during dataset version creation (at least in in my own tests); however, from what I can tell, this options eludes `importer` from being called in `import_metadata_file/task.py` (which is why you are seeing the error message above). Why this may be the case exactly, I am still slightly unclear. I have tried tracing this down but with no success thus far. Perhaps you might have a better understanding of how Geostore utilizes Pystac to shine some light on why this may be? I will continue to explore and will update if I am able to work this out.

The second option `root_catalog.save(catalog_type=CatalogType.SELF_CONTAINED)` is the option I went with in this PR. This appears to only be a partial fix. It prevents `item.json` from being updated with every new dataset version creation; however, it doesn't prevent `collection.json` from being updated every time a new dataset is added or updated.

File: geostore/populate_catalog/task.py (hunk 4 above)
Comment (l0b0) [reply geostore/populate_catalog/task.py:66]: In that case, how about replacing the method body with `raise NotImplementedError`? That way we'll at least know to remove the coverage pragma once it's used again.

**Review by l0b0 — DISMISSED** [review]

**Review by billgeo — APPROVED** [review]

**Review by AmrouEmad — APPROVED** [review]
